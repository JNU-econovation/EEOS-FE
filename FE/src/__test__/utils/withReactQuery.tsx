import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

const queryClient = new QueryClient();

const withReactQuery = (Component: React.ReactNode) => {
  return () => (
    <QueryClientProvider client={queryClient}>{Component}</QueryClientProvider>
  );
};

export default withReactQuery;
