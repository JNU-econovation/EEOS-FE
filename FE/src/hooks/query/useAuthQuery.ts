import { postAdminLogin, postSlackLogin } from "@/apis/auth";
import ERROR_CODE from "@/constants/ERROR_CODE";
import ROUTES from "@/constants/ROUTES";
import {
  deleteTokenInfo,
  setAccessToken,
  setTokenExpiration,
} from "@/utils/authWithStorage";
import { useMutation } from "@tanstack/react-query";
import { useRouter } from "next/navigation";

// TODO: 라우팅 및 토큰 처리 로직은 컴포넌트에서 수행
export const useSlackLoginMutation = () => {
  const router = useRouter();
  return useMutation(
    ["slackLogin"],
    async ({ code, redirect_uri }: { code: string; redirect_uri: string }) =>
      await postSlackLogin(code, redirect_uri),
    {
      onSuccess: (data) => {
        const { accessToken, accessExpiredTime } = data;
        setAccessToken(accessToken);
        setTokenExpiration(accessExpiredTime);
        router.replace(ROUTES.MAIN);
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      onError: (error: any) => {
        const errorCode = error?.response?.data?.code;
        errorCode === ERROR_CODE.AUTH.INVALID_NAME &&
          router.replace(ROUTES.NAME_ERROR);
      },
    },
  );
};

export const useAdminLoginMutation = () => {
  const router = useRouter();
  return useMutation({
    mutationFn: ({ id, password }: { id: string; password: string }) =>
      postAdminLogin(id, password),

    onSuccess: (data) => {
      const { accessToken, accessExpiredTime } = data;
      setAccessToken(accessToken);
      setTokenExpiration(accessExpiredTime);
      router.replace(ROUTES.ADMIN_MAIN);
    },
  });
};

export const useLogoutMutation = () => {
  return { mutate: deleteTokenInfo };
};
