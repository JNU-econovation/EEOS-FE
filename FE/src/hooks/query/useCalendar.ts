import { useQuery } from "@tanstack/react-query";
import { getCalendarEventsOnWeek } from "@/apis/calendar";

export const useGetSimpleCalendarEvents = () => {
  return useQuery({
    queryKey: ["calendar", "simpleEvents"],
    queryFn: () =>
      getCalendarEventsOnWeek({
        year: new Date().getFullYear(),
        month: new Date().getMonth() + 1,
        date: new Date().getDate(),
        duration: 14,
      }),
    staleTime: 1000 * 60 * 5,
    cacheTime: 1000 * 60 * 10,
  });
};
