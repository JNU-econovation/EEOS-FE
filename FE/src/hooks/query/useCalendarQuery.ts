import { useMutation, useQuery } from "@tanstack/react-query";
import {
  deleteCalender,
  getCalendarEventsOnWeek,
  postCalender,
} from "@/apis/calendar";
import API from "@/constants/API";
import { DateFilter, NewCalendar } from "@/types/calendar";

export function useCreateCalendarEventMutation() {
  return useMutation({
    mutationFn: (newCalendar: NewCalendar) => postCalender(newCalendar),
  });
}

export function useDeleteCalendarEventMutation() {
  return useMutation({
    mutationFn: (calendarId: number) => deleteCalender(calendarId),
  });
}

export function useFetchMonthlyCalendarQuery(dateFilter: DateFilter) {
  return useQuery({
    queryKey: [
      API.CALENDAR.FETCH,
      dateFilter.year,
      dateFilter.month,
      dateFilter.date,
      dateFilter.duration,
    ],
    queryFn: () => getCalendarEventsOnWeek(dateFilter),
  });
}
