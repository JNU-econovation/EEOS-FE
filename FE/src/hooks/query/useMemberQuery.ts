import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { MemberActiveStatusInfoDto } from "@/apis/dtos/member.dto";
import {
  deleteMember,
  getMembersByActiveStatus,
  getProgramMembersByActiveStatus,
  getProgramMembersByAttendStatus,
  updateMemberActiveStatus,
} from "@/apis/member";
import API from "@/constants/API";
import { ActiveStatus, ActiveStatusWithAll, AttendMode } from "@/types/member";

export const useGetMemberByActive = (activeStatus: ActiveStatusWithAll) => {
  return useQuery({
    queryKey: [API.MEMBER.LIST, activeStatus],
    queryFn: () => getMembersByActiveStatus(activeStatus),
    staleTime: 1000 * 60 * 5,
  });
};

interface GetProgramMemebersByActive {
  status: ActiveStatusWithAll;
  programId: number;
}
interface GetProgramMemebersByAttend {
  status: AttendMode;
  programId: number;
}
interface UpdateMemberActiveStatus {
  memberId: number;
}

export const useGetProgramMembersByActive = ({
  programId,
  status,
}: GetProgramMemebersByActive) => {
  return useQuery({
    queryKey: [API.MEMBER.ACTIVE_STATUS(programId), status],
    queryFn: () => getProgramMembersByActiveStatus(programId, status),
    staleTime: 1000 * 60 * 5,
  });
};

export const useGetProgramMembersByAttend = ({
  programId,
  status,
}: GetProgramMemebersByAttend) => {
  return useQuery({
    queryKey: [API.MEMBER.ATTEND_STATUS(programId), status],
    queryFn: () => getProgramMembersByAttendStatus(programId, status),
    staleTime: 1000 * 60 * 5,
  });
};

export const useUpdateMemberActiveStatus = ({
  memberId,
}: UpdateMemberActiveStatus) => {
  const queryClient = useQueryClient();
  const data = useMutation<void, Error, { activeStatus: ActiveStatus }>({
    mutationFn: ({ activeStatus }: { activeStatus: ActiveStatus }) =>
      updateMemberActiveStatus(memberId, activeStatus),

    onMutate: ({ activeStatus: newActiveStatus }) => {
      const prevMemberActiveData = queryClient.getQueryData<
        MemberActiveStatusInfoDto[]
      >([API.MEMBER.LIST, "all"]);

      const newMemberActiveData = prevMemberActiveData.map((v) =>
        memberId === v.memberId
          ? {
              activeStatus: newActiveStatus,
              memberId: v.memberId,
              name: v.name,
            }
          : v,
      );

      queryClient.setQueryData([API.MEMBER.LIST, "all"], newMemberActiveData);
    },
    onError: () => {
      queryClient.invalidateQueries([API.MEMBER.LIST]);
    },
  });
  return data;
};

export const useDeleteMember = () => {
  const queryClient = useQueryClient();
  const data = useMutation<void, Error, { memberId: number }>({
    mutationFn: ({ memberId }) => deleteMember(memberId),

    onSettled: () => {
      queryClient.invalidateQueries([API.MEMBER.LIST]);
    },
  });
  return data;
};
