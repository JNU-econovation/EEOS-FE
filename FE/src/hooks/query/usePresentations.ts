import { useQuery } from "@tanstack/react-query";
import { getPresentations } from "@/apis/proxy/github";

const usePresentations = (githubUrl: string) => {
  return useQuery({
    queryKey: ["presentations", githubUrl],
    queryFn: () => getPresentations(githubUrl),
  });
};

export default usePresentations;
