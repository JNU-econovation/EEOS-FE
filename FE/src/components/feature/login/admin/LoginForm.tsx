"use client";

import { FormEvent, useEffect, useState } from "react";
import { useAdminLoginMutation } from "@/hooks/query/useAuthQuery";

const LoginForm = () => {
  const [id, setId] = useState("");
  const [password, setPassword] = useState("");

  const { mutate, isError } = useAdminLoginMutation();

  const loginToAdmin = async (e: FormEvent) => {
    e.preventDefault();
    if (!id || !password) {
      alert("아이디와 비밀번호를 입력해주세요.");
      return;
    }
    mutate({ id, password });
  };

  useEffect(() => {
    setPassword("");
  }, [isError]);

  return (
    <form className="mt-12 flex flex-col gap-3" onSubmit={loginToAdmin}>
      <input
        type="text"
        aria-label="아이디"
        name="id"
        placeholder="아이디"
        className="mx-8 p-2"
        value={id}
        onChange={(e) => setId(e.target.value)}
      />
      <input
        type="password"
        aria-label="비밀번호"
        role="password"
        name="password"
        placeholder="비밀번호"
        className="mx-8 p-2"
        value={password}
        onChange={(e) => setPassword(e.target.value)}
      />
      <button className="mt-4 w-full rounded-lg bg-primary p-2 font-semibold">
        로그인
      </button>
    </form>
  );
};

export default LoginForm;
