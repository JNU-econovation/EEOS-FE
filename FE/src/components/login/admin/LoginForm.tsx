"use client";

import { useState } from "react";
import { useFormStatus } from "react-dom";
import { useAdminLoginMutation } from "@/hooks/query/useAuthQuery";

const LoginForm = () => {
  const [id, setId] = useState("");
  const [password, setPassword] = useState("");
  const { pending } = useFormStatus();
  const { mutate } = useAdminLoginMutation();

  const loginToAdmin = async (e) => {
    e.preventDefault();
    if (!id || !password) {
      alert("아이디와 비밀번호를 입력해주세요.");
      return;
    }
    mutate({ id, password });
  };

  return (
    <form className="mt-12 flex flex-col gap-3" onSubmit={loginToAdmin}>
      <input
        type="text"
        name="id"
        placeholder="id"
        className="mx-8 p-2"
        value={id}
        onChange={(e) => setId(e.target.value)}
      />
      <input
        type="password"
        name="password"
        placeholder="password"
        className="mx-8 p-2"
        value={password}
        onChange={(e) => setPassword(e.target.value)}
      />
      <button
        className="mt-4 w-full rounded-lg bg-primary p-2 font-semibold"
        disabled={pending}
      >
        로그인
      </button>
    </form>
  );
};

export default LoginForm;
