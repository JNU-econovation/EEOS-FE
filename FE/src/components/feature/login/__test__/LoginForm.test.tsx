/**
 * @jest-environment jsdom
 */

import { render, screen, waitFor } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import renderer from "react-test-renderer";
import LoginForm from "../admin/LoginForm";
// eslint-disable-next-line import/order
import withReactQuery from "@/__test__/utils/withReactQuery";
import { useAdminLoginMutation } from "@/hooks/query/useAuthQuery";

// react-dom useFormStatus mocking
jest.mock("react-dom", () => {
  return {
    ...jest.requireActual("react-dom"),
    useFormStatus: jest.fn(),
  };
});

// next.js mocking
jest.mock("next/navigation", () => ({
  useSearchParams: jest.fn(() => ({
    get: jest.fn(() => "code"),
  })),
  useRouter: jest.fn(() => ({
    back: jest.fn(),
    replace: jest.fn(),
  })),
}));

// query mocking
const mutate = jest.fn();

jest.mock("@/hooks/query/useAuthQuery", () => ({
  useAdminLoginMutation: jest.fn(() => ({
    mutate,
    isError: false,
  })),
}));

describe("LoginForm.tsx", () => {
  let mockedAlert: jest.SpyInstance;

  beforeEach(() => {
    mockedAlert = jest.spyOn(window, "alert").mockImplementation(() => {});
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("네트워크가 올바른 환경에서", () => {
    it("올바르게 렌더링 된다", async () => {
      const WrappedComponent = withReactQuery(<LoginForm />);
      const tree = renderer.create(<WrappedComponent />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("초기상태(아무 값도 입력하지 않은 상태)에서 로그인 시 아이디와 비밀번호를 입력해주세요 alert가 뜬다", async () => {
      // arrange
      const WrappedComponent = withReactQuery(<LoginForm />);
      render(<WrappedComponent />);

      // act
      const loginButton = screen.getByRole("button", {
        name: "로그인",
      });

      await userEvent.click(loginButton);

      // assert
      expect(mockedAlert).toHaveBeenCalledWith(
        "아이디와 비밀번호를 입력해주세요.",
      );
    });

    describe("아이디를 입력한 상태에서", () => {
      it("로그인 시 아이디와 비밀번호를 입력해주세요 alert가 뜬다", async () => {
        // Arrange
        const id = "testId";

        const WrappedComponent = withReactQuery(<LoginForm />);
        render(<WrappedComponent />);

        // act
        const loginButton = screen.getByRole("button", {
          name: "로그인",
        });
        const idInput = screen.getByRole("textbox", {
          name: "아이디",
        });

        await userEvent.click(loginButton);
        await userEvent.type(idInput, id);

        expect(mockedAlert).toHaveBeenCalledTimes(1);
      });
      describe("비밀번호를 입력한 상태에서", () => {
        it("아이디 혹은 비밀번호가 올바르지 않은 경우 로그인이 진행되지 않는다", async () => {
          // Arrange
          const id = "notCorrectId";
          const password = "notCorrectPassword";

          const WrappedComponent = withReactQuery(<LoginForm />);
          render(<WrappedComponent />);

          // act
          const idInput = screen.getByRole("textbox", {
            name: "아이디",
          });

          const passwordInput = screen.getByRole("password", {
            name: "비밀번호",
          });

          const loginButton = screen.getByRole("button", {
            name: "로그인",
          });

          await userEvent.type(idInput, id);
          await userEvent.type(passwordInput, password);
          await userEvent.click(loginButton);

          // assert
          expect(mutate).toHaveBeenCalled();
          expect(mutate).toHaveBeenCalledWith({
            id,
            password,
          });
        });
        it("아이디 및 비밀번호가 올바른 경우 로그인이 진행된다", async () => {
          // Arrange
          const id = "correctId";
          const password = "correctPassword";

          const WrappedComponent = withReactQuery(<LoginForm />);
          render(<WrappedComponent />);

          // act
          const loginButton = screen.getByRole("button", {
            name: "로그인",
          });
          const idInput = screen.getByRole("textbox", {
            name: "아이디",
          });
          const passwordInput = screen.getByRole("password", {
            name: "비밀번호",
          });

          await userEvent.type(idInput, id);
          await userEvent.type(passwordInput, password);
          await userEvent.click(loginButton);

          expect(mutate).toHaveBeenCalled();
          expect(mutate).toHaveBeenCalledWith({
            id,
            password,
          });
        });
      });
    });
    describe("비밀번호를 입력한 상태에서", () => {
      it("로그인 시 아이디와 비밀번호를 입력해주세요 alert가 뜬다", async () => {
        // Arrange
        const password = "testPassword";

        const WrappedComponent = withReactQuery(<LoginForm />);
        render(<WrappedComponent />);

        // act
        const loginButton = screen.getByRole("button", {
          name: "로그인",
        });

        const passwordInput = screen.getByRole("password", {
          name: "비밀번호",
        });

        await userEvent.type(passwordInput, password);
        await userEvent.click(loginButton);

        // assert
        expect(mockedAlert).toHaveBeenCalledWith(
          "아이디와 비밀번호를 입력해주세요.",
        );
      });

      describe("아이디를 입력한 상태에서", () => {
        it("아이디 혹은 비밀번호가 올바르지 않은 경우 로그인이 진행되지 않는다", async () => {
          // Arrange
          const id = "uncorrectId";
          const password = "uncorrectPassword";

          const WrappedComponent = withReactQuery(<LoginForm />);
          render(<WrappedComponent />);

          mutate.mockReturnValue({
            isError: true,
          });

          // act
          const loginButton = screen.getByRole("button", {
            name: "로그인",
          });
          const idInput = screen.getByRole("textbox", {
            name: "아이디",
          });
          const passwordInput = screen.getByRole("password", {
            name: "비밀번호",
          });

          await userEvent.type(idInput, id);
          await userEvent.type(passwordInput, password);
          await userEvent.click(loginButton);

          // assert
          expect(mutate).toHaveBeenCalled();
          expect(mutate).toHaveBeenCalledWith({
            id,
            password,
          });
        });

        it("아이디 및 비밀번호가 올바른 경우 로그인이 진행된다", async () => {
          // Arrange
          const id = "testId";
          const password = "testPassword";
          const WrappedComponent = withReactQuery(<LoginForm />);
          render(<WrappedComponent />);
          // act
          const loginButton = screen.getByRole("button", {
            name: "로그인",
          });
          const idInput = screen.getByRole("textbox", {
            name: "아이디",
          });
          const passwordInput = screen.getByRole("password", {
            name: "비밀번호",
          });

          await userEvent.type(idInput, id);
          await userEvent.type(passwordInput, password);
          await userEvent.click(loginButton);

          // assert
          expect(mutate).toHaveBeenCalled();
          expect(mutate).toHaveBeenCalledWith({
            id,
            password,
          });
        });
      });
    });
  });

  describe("네트워크가 올바르지 않은 환경에서", () => {
    it("로그인 시 에러가 발생하면 비밀번호를 초기화한다", async () => {
      // Arrange
      const WrappedComponent = withReactQuery(<LoginForm />);
      render(<WrappedComponent />);
      const loginButton = screen.getByRole("button", {
        name: "로그인",
      });
      const idInput = screen.getByRole("textbox", {
        name: "아이디",
      });
      const passwordInput = screen.getByRole("password", {
        name: "비밀번호",
      }) as HTMLInputElement;

      useAdminLoginMutation.mockReturnValue({
        mutate,
        isError: true,
      });

      // act
      await userEvent.type(idInput, "testId");
      await userEvent.type(passwordInput, "testPassword");

      await userEvent.click(loginButton);

      // assert
      expect(mutate).toHaveBeenCalled();
      waitFor(() => {
        expect(passwordInput.value).toBe("");
      });
    });
  });
});
