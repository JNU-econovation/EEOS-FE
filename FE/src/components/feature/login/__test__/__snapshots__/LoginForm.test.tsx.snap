// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm 네트워크가 올바른 환경에서 올바르게 렌더링 된다 1`] = `
<form
  className="mt-12 flex flex-col gap-3"
  onSubmit={[Function]}
>
  <input
    aria-label="아이디"
    className="mx-8 p-2"
    name="id"
    onChange={[Function]}
    placeholder="아이디"
    type="text"
    value=""
  />
  <input
    aria-label="비밀번호"
    className="mx-8 p-2"
    name="password"
    onChange={[Function]}
    placeholder="비밀번호"
    role="password"
    type="password"
    value=""
  />
  <button
    className="mt-4 w-full rounded-lg bg-primary p-2 font-semibold"
  >
    로그인
  </button>
</form>
`;

exports[`LoginForm 올바르게 렌더링 된다 1`] = `
<form
  className="mt-12 flex flex-col gap-3"
  onSubmit={[Function]}
>
  <input
    aria-label="아이디"
    className="mx-8 p-2"
    name="id"
    onChange={[Function]}
    placeholder="아이디"
    type="text"
    value=""
  />
  <input
    aria-label="비밀번호"
    className="mx-8 p-2"
    name="password"
    onChange={[Function]}
    placeholder="비밀번호"
    role="password"
    type="password"
    value=""
  />
  <button
    className="mt-4 w-full rounded-lg bg-primary p-2 font-semibold"
  >
    로그인
  </button>
</form>
`;

exports[`LoginForm.tsx 네트워크가 올바른 환경에서 올바르게 렌더링 된다 1`] = `
<form
  className="mt-12 flex flex-col gap-3"
  onSubmit={[Function]}
>
  <input
    aria-label="아이디"
    className="mx-8 p-2"
    name="id"
    onChange={[Function]}
    placeholder="아이디"
    type="text"
    value=""
  />
  <input
    aria-label="비밀번호"
    className="mx-8 p-2"
    name="password"
    onChange={[Function]}
    placeholder="비밀번호"
    role="password"
    type="password"
    value=""
  />
  <button
    className="mt-4 w-full rounded-lg bg-primary p-2 font-semibold"
  >
    로그인
  </button>
</form>
`;
